docs/man/btdownloadcurses.bittornado.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.33.
.TH BTDOWNLOADCURSES "1" "August 2004" "btdownloadcurses (bittornado)" "User Commands"
.SH NAME
Btdownloadcurses \- curses bittornado download interface
.SH SYNOPSIS
.B btdownloadcurses
\fI<global options>\fR
.SH DESCRIPTION
Curses interface to download torrents.
.SH OPTIONS
This program follows the usual GNU command-line syntax, with long options
starting with two dashes ('-'). A summary of options is included below.

.PP
\fB\-\-max_uploads\fR \fInumber\fP
.IP
the maximum \fInumber\fP of uploads to allow at once. (defaults to 7)
.PP
\fB\-\-keepalive_interval\fR \fIseconds\fP
.IP
number of \fIseconds\fP to pause between sending keepalives (defaults to 120.0)
.PP
\fB\-\-download_slice_size\fR \fIbytes\fP
.IP
How many \fIbytes\fP to query for per request. (defaults to 16384)
.PP
\fB\-\-upload_unit_size\fR \fIbytes\fP
.IP
when limiting upload rate, how many \fIbytes\fP to send at a time (defaults to 1460)
.PP
\fB\-\-request_backlog\fR \fInumber\fP
.IP
maximum \fInumber\fP of requests to keep in a single pipe at once. (defaults to 10)
.PP
\fB\-\-max_message_length\fR \fIlength\fP
.IP
maximum \fIlength\fP prefix encoding you'll accept over the wire - larger values get the
connection dropped. (defaults to 8388608)
.PP
\fB\-\-ip\fR \fIip\fP
.IP
\fIip\fP to report you have to the tracker. (defaults to '')
.PP
\fB\-\-minport\fR \fIportnum\fP
.IP
set \fIportnum\fP as the minimum port to listen on, counts up if unavailable (defaults to 10000)
.PP
\fB\-\-maxport\fR \fIportnum\fP
.IP
set \fIportnum\fP as the maximum port to listen on (defaults to 60000)
.PP
\fB\-\-random_port\fR \fI 0 | 1 \fP
.IP
whether to choose randomly inside the port range instead of counting up linearly
(defaults to 1)
.PP
\fB\-\-responsefile\fR \fIfile\fP
.IP
\fIfile\fP the server response was stored in, alternative to url (defaults to '')
.PP
\fB\-\-url\fR \fIURL\fP
.IP
\fIURL\fP to get file from, alternative to responsefile (defaults to '')
.PP
\fB\-\-selector_enabled\fR \fI 0 | 1 \fP
.IP
whether to enable the file selector and fast resume function (defaults to 1)
.PP
\fB\-\-expire_cache_data\fR \fIdays\fP
.IP
the number of \fIdays\fP after which you wish to expire old cache data (0 = disabled) (defaults
to 10)
.PP
\fB\-\-priority\fR \fI -1|0|1|2[,-1|0|1|2] \fP
.IP
a list of file priorities separated by commas, must be one per file, 0 = highest, 1 =
normal, 2 = lowest, \fB\-1\fR = download disabled (defaults to ''). Order is based
on the file/torrent order as shown by btshowmetainfo. For example, to download only
the third of four files use: --priority -1,-1,2,-1
.PP
\fB\-\-saveas\fR \fIfilename\fP
.IP
local \fIfilename\fP to save the file as, null indicates query user (defaults to '')
.PP
\fB\-\-timeout\fR \fIseconds\fP
.IP
time in \fIseconds\fP to wait between closing sockets which nothing has been received on (defaults to
300.0)
.PP
\fB\-\-timeout_check_interval\fR \fIseconds\fP
.IP
time to wait in \fIseconds\fP between checking if any connections have timed out (defaults to 60.0)
.PP
\fB\-\-max_slice_length\fR \fIlength\fP
.IP
maximum \fIlength\fP slice to send to peers, larger requests are ignored (defaults to 131072)
.PP
\fB\-\-max_rate_period\fR \fIseconds\fP
.IP
maximum amount of time in \fIseconds\fP to guess the current rate estimate represents (defaults to 20.0)
.PP
\fB\-\-bind\fR \fIip[,hostname]\fP
.IP
comma-separated list of \fIips/hostnames\fP to bind to locally (defaults to '')
.PP
\fB\-\-ipv6_enabled\fR \fI 0 | 1 \fP
.IP
allow the client to connect to peers via IPv6 (defaults to 0)
.PP
\fB\-\-ipv6_binds_v4\fR \fI 0 | 1 \fP
.IP
set if an IPv6 server socket will also field IPv4 connections (defaults to 1)
.PP
\fB\-\-upnp_nat_access\fR \fI 0 | 1 | 2 \fP
.IP
attempt to autoconfigure a UPnP router to forward a server port (0 = disabled, 1 = mode 1
[fast], 2 = mode 2 [slow]) (defaults to 1)
.PP
\fB\-\-upload_rate_fudge\fR \fIseconds\fP
.IP
time equivalent in \fIseconds\fP of writing to kernel-level TCP buffer, for rate adjustment (defaults to
5.0)
.PP
\fB\-\-tcp_ack_fudge\fR \fIoverhead\fP
.IP
how much TCP ACK download \fIoverhead\fP to add to upload rate calculations (0 = disabled)
(defaults to 0.029999999999999999)
.PP
\fB\-\-display_interval\fR \fIseconds\fP
.IP
time in \fIseconds\fP between updates of displayed information (defaults to 0.5)
.PP
\fB\-\-rerequest_interval\fR \fIseconds\fP
.IP
time to wait, in \fIseconds\fP, between requesting more peers (defaults to 300)
.PP
\fB\-\-min_peers\fR \fInumber\fP
.IP
minimum \fInumber\fP of peers to not do rerequesting (defaults to 20)
.PP
\fB\-\-http_timeout\fR \fIseconds\fP
.IP
number of \fIseconds\fP to wait before assuming that an http connection has timed out (defaults
to 60)
.PP
\fB\-\-max_initiate\fR \fInumber\fP
.IP
\fInumber\fP of peers at which to stop initiating new connections (defaults to 40)
.PP
\fB\-\-check_hashes\fR \fI 0 | 1 \fP
.IP
whether to check hashes on disk (defaults to 1)
.PP
\fB\-\-max_upload_rate\fR \fIkB/s\fP
.IP
maximum \fIkB/s\fP to upload at (0 = no limit, \fB\-1\fR = automatic) (defaults to 0)
.PP
\fB\-\-max_download_rate\fR \fIkB/s\fP
.IP
maximum \fIkB/s\fP to download at (0 = no limit) (defaults to 0)
.PP
\fB\-\-alloc_type\fR \fInormal | background | pre-allocate | sparse\fP
.IP
allocation type (may be normal, background, pre-allocate or sparse) (defaults to
\&'normal')
.PP
\fB\-\-alloc_rate\fR \fIMiB/s\fP
.IP
rate (in \fIMiB/s\fP) to allocate space at using background allocation (defaults to 2.0)
.PP
\fB\-\-buffer_reads\fR \fI 0 | 1 \fP
.IP
whether to buffer disk reads (defaults to 1)
.PP
\fB\-\-write_buffer_size\fR \fIspace\fP
.IP
the maximum amount of \fIspace\fP to use for buffering disk writes (in megabytes, 0 = disabled)
(defaults to 4)
.PP
\fB\-\-snub_time\fR \fIseconds\fP
.IP
\fIseconds\fP to wait for data to come in over a connection before assuming it's
semi-permanently choked (defaults to 30.0)
.PP
\fB\-\-spew\fR \fI 0 | 1 \fP
.IP
whether to display diagnostic info to stdout (defaults to 0)
.PP
\fB\-\-rarest_first_cutoff\fR \fInumber\fP
.IP
\fInumber\fP of downloads at which to switch from random to rarest first (defaults to 2)
.PP
\fB\-\-rarest_first_priority_cutoff\fR \fInumber\fP
.IP
the \fInumber\fP of peers which need to have a piece before other partials take priority over
rarest first (defaults to 5)
.PP
\fB\-\-min_uploads\fR \fInumber\fP
.IP
the \fInumber\fP of uploads to fill out to with extra optimistic unchokes (defaults to 4)
.PP
\fB\-\-max_files_open\fR \fInumber\fP
.IP
the maximum \fInumber\fP of files to keep open at a time, 0 means no limit (defaults to 50)
.PP
\fB\-\-round_robin_period\fR \fIseconds\fP
.IP
the number of \fIseconds\fP between the client's switching upload targets (defaults to 30)
.PP
\fB\-\-super_seeder\fR \fI 0 | 1 \fP
.IP
whether to use special upload-efficiency-maximizing routines (only for dedicated seeds)
(defaults to 0)
.PP
\fB\-\-security\fR \fI 0 | 1 \fP
.IP
whether to enable extra security features intended to prevent abuse (defaults to 1)
.PP
\fB\-\-max_connections\fR \fInumber\fP
.IP
the absolute maximum \fInumber\fP of peers to connect with (0 = no limit) (defaults to 0)
.PP
\fB\-\-auto_kick\fR \fI 0 | 1 \fP
.IP
whether to allow the client to automatically kick/ban peers that send bad data (defaults
to 1)
.PP
\fB\-\-double_check\fR \fI 0 | 1 \fP
.IP
whether to double-check data being written to the disk for errors (may increase CPU load)
(defaults to 1)
.PP
\fB\-\-triple_check\fR \fI 0 | 1 \fP
.IP
whether to thoroughly check data being written to the disk (may slow disk access)
(defaults to 0)
.PP
\fB\-\-lock_files\fR \fI 0 | 1 \fP
.IP
whether to lock files the client is working with (defaults to 1)
.PP
\fB\-\-lock_while_reading\fR \fI 0 | 1 \fP
.IP
whether to lock access to files being read (defaults to 0)
.PP
\fB\-\-auto_flush\fR \fIminutes\fP
.IP
\fIminutes\fP between automatic flushes to disk (0 = disabled) (defaults to 0)
.PP
.SH SEE ALSO
.BR bittorrent-downloader (1),
.BR bittorrent-multi-downloader (1),
.BR btdownloadheadless (1).
.br
.SH AUTHOR
This manual page was written by Micah Anderson <micah@riseup.net>,
for the Debian GNU/Linux system (but may be used by others).


btdownloadcurses.py
#!/usr/bin/env python

# Written by Henry 'Pi' James
# see LICENSE.txt for license information

SPEW_SCROLL_RATE = 1

import sys
import os
import time
import signal
import random
import socket
import hashlib
import threading
from BitTornado.Client.download_bt1 import BT1Download, defaults, \
    parse_params, get_usage, get_response
from BitTornado.Network.RawServer import RawServer
from BitTornado.Network.SocketHandler import UPnP_ERROR
from BitTornado.Meta.bencode import bencode
from BitTornado.Network.natpunch import UPnP_test
from BitTornado.clock import clock
from BitTornado import version
from BitTornado.Application.ConfigDir import ConfigDir
from BitTornado.Application.PeerID import createPeerID

try:
    import curses
    import curses.panel
    from curses.wrapper import wrapper as curses_wrapper
except ImportError:
    print 'Textmode GUI initialization failed, cannot proceed.'
    print
    print 'This download interface requires the standard Python module ' \
          '"curses", which is unfortunately not available for the native ' \
          'Windows port of Python. It is however available for the Cygwin ' \
          'port of Python, running on all Win32 systems (www.cygwin.com).'
    print
    print 'You may still use "btdownloadheadless.py" to download.'
    sys.exit(1)


def fmttime(n):
    if n == 0:
        return 'download complete!'
    try:
        n = int(n)
        assert n >= 0 and n < 5184000  # 60 days
    except (AssertionError, ValueError):
        return '<unknown>'
    m, s = divmod(n, 60)
    h, m = divmod(m, 60)
    return 'finishing in %d:%02d:%02d' % (h, m, s)


def fmtsize(n):
    s = str(n)
    size = s[-3:]
    while len(s) > 3:
        s = s[:-3]
        size = '%s,%s' % (s[-3:], size)
    if n > 999:
        unit = ['B', 'KiB', 'MiB', 'GiB', 'TiB', 'PiB', 'EiB', 'ZiB', 'YiB']
        i = 1
        while i + 1 < len(unit) and (n >> 10) >= 999:
            i += 1
            n >>= 10
        n = float(n) / (1 << 10)
        size = '%s (%.0f %s)' % (size, n, unit[i])
    return size


class CursesDisplayer:
    def __init__(self, scrwin, errlist, doneflag):
        self.scrwin = scrwin
        self.errlist = errlist
        self.doneflag = doneflag

        signal.signal(signal.SIGWINCH, self.winch_handler)
        self.changeflag = threading.Event()

        self.done = 0
        self.file = ''
        self.fileSize = ''
        self.activity = ''
        self.status = ''
        self.progress = ''
        self.downloadTo = ''
        self.downRate = '---'
        self.upRate = '---'
        self.shareRating = ''
        self.seedStatus = ''
        self.peerStatus = ''
        self.errors = []
        self.last_update_time = 0
        self.spew_scroll_time = 0
        self.spew_scroll_pos = 0

        self._remake_window()

    def winch_handler(self, signum, stackframe):
        self.changeflag.set()
        curses.endwin()
        self.scrwin.refresh()
        self.scrwin = curses.newwin(0, 0, 0, 0)
        self._remake_window()

    def _remake_window(self):
        self.scrh, self.scrw = self.scrwin.getmaxyx()
        self.scrpan = curses.panel.new_panel(self.scrwin)
        self.labelh, self.labelw, self.labely, self.labelx = 11, 9, 1, 2
        self.labelwin = curses.newwin(self.labelh, self.labelw,
                                      self.labely, self.labelx)
        self.labelpan = curses.panel.new_panel(self.labelwin)
        self.fieldh, self.fieldw, self.fieldy, self.fieldx = (
            self.labelh, (self.scrw - 2) - (self.labelw - 3), 1,
            self.labelw + 3)
        self.fieldwin = curses.newwin(self.fieldh, self.fieldw,
                                      self.fieldy, self.fieldx)
        self.fieldwin.nodelay(1)
        self.fieldpan = curses.panel.new_panel(self.fieldwin)
        self.spewh, self.speww, self.spewy, self.spewx = (
            self.scrh - self.labelh - 2, self.scrw - 3, 1 + self.labelh, 2)
        self.spewwin = curses.newwin(self.spewh, self.speww,
                                     self.spewy, self.spewx)
        self.spewpan = curses.panel.new_panel(self.spewwin)
        try:
            self.scrwin.border(*map(ord, '||--    '))
        except Exception:
            pass
        self.labelwin.addstr(0, 0, 'file:')
        self.labelwin.addstr(1, 0, 'size:')
        self.labelwin.addstr(2, 0, 'dest:')
        self.labelwin.addstr(3, 0, 'progress:')
        self.labelwin.addstr(4, 0, 'status:')
        self.labelwin.addstr(5, 0, 'dl speed:')
        self.labelwin.addstr(6, 0, 'ul speed:')
        self.labelwin.addstr(7, 0, 'sharing:')
        self.labelwin.addstr(8, 0, 'seeds:')
        self.labelwin.addstr(9, 0, 'peers:')
        curses.panel.update_panels()
        curses.doupdate()
        self.changeflag.clear()

    def finished(self):
        self.done = 1
        self.activity = 'download succeeded!'
        self.downRate = '---'
        self.display(fractionDone=1)

    def failed(self):
        self.done = 1
        self.activity = 'download failed!'
        self.downRate = '---'
        self.display()

    def error(self, errormsg):
        newerrmsg = time.strftime('[%H:%M:%S] ') + errormsg
        self.errors.append(newerrmsg)
        self.errlist.append(newerrmsg)
        self.display()

    def display(self, dpflag=threading.Event(), fractionDone=None,
                timeEst=None, downRate=None, upRate=None, activity=None,
                statistics=None, spew=None, **kwargs):

        inchar = self.fieldwin.getch()
        if inchar == 12:                    # ^L
            self._remake_window()
        elif inchar in (ord('q'), ord('Q')):
            self.doneflag.set()

        if activity is not None and not self.done:
            self.activity = activity
        elif timeEst is not None:
            self.activity = fmttime(timeEst)
        if self.changeflag.isSet() or \
                self.last_update_time + 0.1 > clock() and \
                fractionDone not in (0.0, 1.0) and \
                activity is not None:
            return
        self.last_update_time = clock()
        if fractionDone is not None:
            blocknum = int(self.fieldw * fractionDone)
            self.progress = blocknum * '#' + (self.fieldw - blocknum) * '_'
            self.status = '%s (%.1f%%)' % (self.activity, fractionDone * 100)
        else:
            self.status = self.activity
        if downRate is not None:
            self.downRate = '%.1f KB/s' % (float(downRate) / (1 << 10))
        if upRate is not None:
            self.upRate = '%.1f KB/s' % (float(upRate) / (1 << 10))
        if statistics is not None:
            if statistics.shareRating < 0 or statistics.shareRating > 100:
                shareRateString = "oo"
            else:
                shareRateString = '{:.3f}'.format(statistics.shareRating)
            self.shareRating = '{}  ({:.1f} MB up / {:.1f} MB down)'.format(
                shareRateString,
                float(statistics.upTotal) / (1 << 20),
                float(statistics.downTotal) / (1 << 20))
            if self.done:
                seeds = '{:d} seen recently, '.format(statistics.numOldSeeds)
                copies = 'plus {:.3f} distributed copies'.format(
                    round(statistics.numCopies, 3))
            else:
                seeds = '{:d} seen now, '.format(statistics.numSeeds)
                copies = 'plus {:.3f} distributed copies'.format(
                    round(statistics.numCopies2, 3))
            self.seedStatus = seeds + copies
            self.peerStatus = '{:d} seen now, {:.1f}% done at {:.1f} kB/s' \
                ''.format(statistics.numPeers, statistics.percentDone,
                          float(statistics.torrentRate) / (1 << 10))

        self.fieldwin.erase()
        self.fieldwin.addnstr(0, 0, self.file, self.fieldw, curses.A_BOLD)
        self.fieldwin.addnstr(1, 0, self.fileSize, self.fieldw)
        self.fieldwin.addnstr(2, 0, self.downloadTo, self.fieldw)
        if self.progress:
            self.fieldwin.addnstr(3, 0, self.progress, self.fieldw,
                                  curses.A_BOLD)
        self.fieldwin.addnstr(4, 0, self.status, self.fieldw)
        self.fieldwin.addnstr(5, 0, self.downRate, self.fieldw)
        self.fieldwin.addnstr(6, 0, self.upRate, self.fieldw)
        self.fieldwin.addnstr(7, 0, self.shareRating, self.fieldw)
        self.fieldwin.addnstr(8, 0, self.seedStatus, self.fieldw)
        self.fieldwin.addnstr(9, 0, self.peerStatus, self.fieldw)

        self.spewwin.erase()

        if not spew:
            errsize = self.spewh
            if self.errors:
                self.spewwin.addnstr(0, 0, "error(s):", self.speww,
                                     curses.A_BOLD)
                errsize = len(self.errors)
                displaysize = min(errsize, self.spewh)
                displaytop = errsize - displaysize
                for i in range(displaysize):
                    self.spewwin.addnstr(
                        i, self.labelw, self.errors[displaytop + i],
                        self.speww - self.labelw - 1, curses.A_BOLD)
        else:
            if self.errors:
                self.spewwin.addnstr(0, 0, "error:", self.speww, curses.A_BOLD)
                self.spewwin.addnstr(
                    0, self.labelw, self.errors[-1],
                    self.speww - self.labelw - 1, curses.A_BOLD)
            self.spewwin.addnstr(
                2, 0, '  #     IP                 Upload           Download  '
                '   Completed  Speed', self.speww, curses.A_BOLD)

            if self.spew_scroll_time + SPEW_SCROLL_RATE < clock():
                self.spew_scroll_time = clock()
                if len(spew) > self.spewh - 5 or self.spew_scroll_pos > 0:
                    self.spew_scroll_pos += 1
            if self.spew_scroll_pos > len(spew):
                self.spew_scroll_pos = 0

            for i, subspew in enumerate(spew, 1):
                subspew['lineno'] = i
            spew.append({'lineno': None})
            spew = spew[self.spew_scroll_pos:] + spew[:self.spew_scroll_pos]

            for i in range(min(self.spewh - 5, len(spew))):
                if not spew[i]['lineno']:
                    continue
                self.spewwin.addnstr(i + 3, 0, '%3d' % spew[i]['lineno'], 3)
                self.spewwin.addnstr(i + 3, 4,
                                     spew[i]['ip'] + spew[i]['direction'], 16)
                if spew[i]['uprate'] > 100:
                    self.spewwin.addnstr(i + 3, 20, '{:6.0f} KB/s'.format(
                        float(spew[i]['uprate']) / 1000), 11)
                self.spewwin.addnstr(i + 3, 32, '-----', 5)
                if spew[i]['uinterested']:
                    self.spewwin.addnstr(i + 3, 33, 'I', 1)
                if spew[i]['uchoked']:
                    self.spewwin.addnstr(i + 3, 35, 'C', 1)
                if spew[i]['downrate'] > 100:
                    self.spewwin.addnstr(i + 3, 38, '{:6.0f} KB/s'.format(
                        float(spew[i]['downrate']) / 1000), 11)
                self.spewwin.addnstr(i + 3, 50, '-------', 7)
                if spew[i]['dinterested']:
                    self.spewwin.addnstr(i + 3, 51, 'I', 1)
                if spew[i]['dchoked']:
                    self.spewwin.addnstr(i + 3, 53, 'C', 1)
                if spew[i]['snubbed'] == 1:
                    self.spewwin.addnstr(i + 3, 55, 'S', 1)
                self.spewwin.addnstr(i + 3, 58, '{:6.1%}'.format(
                    spew[i]['completed']), 6)
                if spew[i]['speed'] is not None:
                    self.spewwin.addnstr(i + 3, 64, '{:5.0f} KB/s'.format(
                        float(spew[i]['speed']) / 1000), 10)

            if statistics is not None:
                self.spewwin.addnstr(
                    self.spewh - 1, 0, 'downloading {:d} pieces, have %d '
                    'fragments, %d of %d pieces completed'.format(
                        statistics.storage_active,
                        statistics.storage_dirty,
                        statistics.storage_numcomplete,
                        statistics.storage_totalpieces),
                    self.speww - 1)

        curses.panel.update_panels()
        curses.doupdate()
        dpflag.set()

    def chooseFile(self, default, size, saveas, isdir):
        self.file = default
        self.fileSize = fmtsize(size)
        if saveas == '':
            saveas = default
        self.downloadTo = os.path.abspath(saveas)
        return saveas


def run(scrwin, errlist, params):
    doneflag = threading.Event()
    d = CursesDisplayer(scrwin, errlist, doneflag)
    try:
        while 1:
            configdir = ConfigDir('downloadcurses')
            defaultsToIgnore = ['responsefile', 'url', 'priority']
            configdir.setDefaults(defaults, defaultsToIgnore)
            configdefaults = configdir.loadConfig()
            defaults.append(
                ('save_options', 0, 'whether to save the current options as '
                 'the new default configuration (only for btdownloadcurses.py)'
                 ))
            try:
                config = parse_params(params, configdefaults)
            except ValueError as e:
                d.error('error: {}\nrun with no args for parameter '
                        'explanations'.format(e))
                break
            if not config:
                d.error(get_usage(defaults, d.fieldw, configdefaults))
                break
            if config['save_options']:
                configdir.saveConfig(config)
            configdir.deleteOldCacheData(config['expire_cache_data'])

            myid = createPeerID()
            random.seed(myid)

            rawserver = RawServer(
                doneflag, config['timeout_check_interval'], config['timeout'],
                ipv6_enable=config['ipv6_enabled'], failfunc=d.failed,
                errorfunc=d.error)

            upnp_type = UPnP_test(config['upnp_nat_access'])
            while True:
                try:
                    listen_port = rawserver.find_and_bind(
                        config['minport'], config['maxport'], config['bind'],
                        ipv6_socket_style=config['ipv6_binds_v4'],
                        upnp=upnp_type, randomizer=config['random_port'])
                    break
                except socket.error as e:
                    if upnp_type and e == UPnP_ERROR:
                        d.error('WARNING: COULD NOT FORWARD VIA UPnP')
                        upnp_type = 0
                        continue
                    d.error("Couldn't listen - " + str(e))
                    d.failed()
                    return

            response = get_response(config['responsefile'], config['url'],
                                    d.error)
            if not response:
                break

            infohash = hashlib.sha1(bencode(response['info'])).digest()

            dow = BT1Download(
                d.display, d.finished, d.error, d.error, doneflag, config,
                response, infohash, myid, rawserver, listen_port, configdir)

            if not dow.saveAs(d.chooseFile):
                break

            if not dow.initFiles(old_style=True):
                break
            if not dow.startEngine():
                dow.shutdown()
                break
            dow.startRerequester()
            dow.autoStats()

            if not dow.am_I_finished():
                d.display(activity='connecting to peers')
            rawserver.listen_forever(dow.getPortHandler())
            d.display(activity='shutting down')
            dow.shutdown()
            break

    except KeyboardInterrupt:
        # ^C to exit...
        pass
    try:
        rawserver.shutdown()
    except Exception:
        pass
    if not d.done:
        d.failed()


if __name__ == '__main__':
    if sys.argv[1:] == ['--version']:
        print version
        sys.exit(0)
    if len(sys.argv) <= 1:
        print "Usage: btdownloadcurses.py <global options>\n"
        print get_usage(defaults)
        sys.exit(1)

    errlist = []
    curses_wrapper(run, errlist, sys.argv[1:])

    if errlist:
        print "These errors occurred during execution:"
        for error in errlist:
            print error
